src/vm/runners/builtin_runner/hash.rs

```diff
@@ -115,9 +115,31 @@ impl HashBuiltinRunner {
     }
 
     pub fn get_allocated_memory_units(&self, vm: &VirtualMachine) -> Result<usize, MemoryError> {
-        let value = safe_div_usize(vm.current_step, self.ratio as usize)
-            .map_err(|_| MemoryError::ErrorCalculatingMemoryUnits)?;
-        Ok(self.cells_per_instance as usize * value)
+        match self.ratio {
+            None => {
+                // Dynamic layout has the exact number of instances it needs (up to a power of 2).
+                let instances: usize =
+                    self.get_used_cells(&vm.segments)? / self.cells_per_instance as usize;
+                let components =
+                    (instances / self.instances_per_component as usize).next_power_of_two();
+                Ok(self.cells_per_instance as usize
+                    * self.instances_per_component as usize
+                    * components)
+            }
+            Some(ratio) => {
+                let min_step = (ratio * self.instances_per_component) as usize;
+                if vm.current_step < min_step {
+                    return Err(InsufficientAllocatedCellsError::MinStepNotReached(
+                        min_step,
+                        HASH_BUILTIN_NAME,
+                    )
+                    .into());
+                };
+                let value = safe_div_usize(vm.current_step, ratio as usize)
+                    .map_err(|_| MemoryError::ErrorCalculatingMemoryUnits)?;
+                Ok(self.cells_per_instance as usize * value)
+            }
+        }
```

Comment:
This logic seems shared by all builtins. It may be a good idea to extract it to a separate function.

Reply:
We can move it into the`BuiltinRunner` implementation along with `get_used_cells_and_allocated_size`

Reply:
I take this back, we cant do so without exposing the fields of each builtin

Reply:
Not even adding a method to retrieve it?

Reply:
We can add methods to expose `cells_per_instance` and `instances_per_component` and move the implementations of `get_allocated_memory_units` and `get_used_cells_and_allocated_size` to the `BuiltinRunner` implementation but wouldn't that be exposing too much of the builtin internals?

Reply:
Sounds good